File: frontend/src/lib/api/portfolio.ts
import { apiFetch } from '@/lib/api-client';
export interface PortfolioItem {
  id: string;
  title: string;
  slug: string;
  description: string | null;
  cover_image: string | null;
  demo_url: string | null;
  github_url: string | null;
  technologies: string[] | null;
  start_date: string | null;
  end_date: string | null;
  /** completed=已完成 in_progress=进行中 planned=规划中 */
  status: string | null;
  is_featured: boolean | null;
  sort_order: number | null;
  created_at: string;
  updated_at?: string;
}

export interface PortfolioItemCreate {
  title: string;
  slug: string;
  description?: string | null;
  cover_image?: string | null;
  demo_url?: string | null;
  github_url?: string | null;
  technologies?: string[] | null;
  start_date?: string | null;
  end_date?: string | null;
  status?: string | null;
  is_featured?: boolean | null;
  sort_order?: number | null;
}

export interface PortfolioItemUpdate {
  title?: string;
  slug?: string;
  description?: string | null;
  cover_image?: string | null;
  demo_url?: string | null;
  github_url?: string | null;
  technologies?: string[] | null;
  start_date?: string | null;
  end_date?: string | null;
  status?: string | null;
  is_featured?: boolean | null;
  sort_order?: number | null;
}

export const getPortfolioItems = async (params?: {
  skip?: number;
  limit?: number;
  is_active?: boolean;
}): Promise<PortfolioItem[]> => {
  const queryParams = new URLSearchParams();
  if (params?.skip !== undefined) {queryParams.append('skip', params.skip.toString());}
  if (params?.limit !== undefined) {queryParams.append('limit', params.limit.toString());}
  if (params?.is_active !== undefined) {queryParams.append('is_active', params.is_active.toString());}

  const response = await apiFetch(`/portfolio/?${queryParams.toString()}`, {
    headers: {
      'Content-Type': 'application/json',
    },
  });

  if (!response.ok) {
    const errorData = await response.json().catch(() => ({}));
    throw new Error(errorData.detail || `请求失败: ${response.status}`);
  }

  return response.json();
};

export const getPortfolioItemById = async (id: string): Promise<PortfolioItem> => {
  const response = await apiFetch(`/portfolio/${id}`, {
    headers: {
      'Content-Type': 'application/json',
    },
  });

  if (!response.ok) {
    if (response.status === 404) {
      throw new Error('作品集项目不存在');
    }
    const errorData = await response.json().catch(() => ({}));
    throw new Error(errorData.detail || `请求失败: ${response.status}`);
  }

  return response.json();
};

export const createPortfolioItem = async (portfolioItem: PortfolioItemCreate): Promise<PortfolioItem> => {
  const token = typeof window !== 'undefined' ? localStorage.getItem('auth_token') : null;

  const response = await apiFetch(`/portfolio/`, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      ...(token && { 'Authorization': `Bearer ${token}` }),
    },
    body: JSON.stringify(portfolioItem),
  });

  if (!response.ok) {
    const errorData = await response.json().catch(() => ({}));
    throw new Error(errorData.detail || `请求失败: ${response.status}`);
  }

  return response.json();
};

export const updatePortfolioItem = async (id: string, portfolioItem: PortfolioItemUpdate): Promise<PortfolioItem> => {
  const token = typeof window !== 'undefined' ? localStorage.getItem('auth_token') : null;

  const response = await apiFetch(`/portfolio/${id}`, {
    method: 'PUT',
    headers: {
      'Content-Type': 'application/json',
      ...(token && { 'Authorization': `Bearer ${token}` }),
    },
    body: JSON.stringify(portfolioItem),
  });

  if (!response.ok) {
    const errorData = await response.json().catch(() => ({}));
    throw new Error(errorData.detail || `请求失败: ${response.status}`);
  }

  return response.json();
};

export const deletePortfolioItem = async (id: string): Promise<void> => {
  const token = typeof window !== 'undefined' ? localStorage.getItem('auth_token') : null;

  const response = await apiFetch(`/portfolio/${id}`, {
    method: 'DELETE',
    headers: {
      'Content-Type': 'application/json',
      ...(token && { 'Authorization': `Bearer ${token}` }),
    },
  });

  if (!response.ok) {
    const errorData = await response.json().catch(() => ({}));
    throw new Error(errorData.detail || `请求失败: ${response.status}`);
  }
};

File: backend/app/schemas/portfolio.py
from pydantic import BaseModel, field_serializer, field_validator
from typing import Optional, List
from datetime import date, datetime
from uuid import UUID
import json


# PortfolioItem 相关的类定义（保持向后兼容）

class PortfolioItemBase(BaseModel):
    title: str
    slug: str
    description: Optional[str] = None
    cover_image: Optional[str] = None
    demo_url: Optional[str] = None
    github_url: Optional[str] = None
    technologies: Optional[List[str]] = []  # 使用JSON格式存储技术栈
    start_date: Optional[date] = None
    end_date: Optional[date] = None
    status: Optional[str] = 'completed'  # completed, in_progress, planned
    is_featured: Optional[bool] = False
    sort_order: Optional[int] = 0

    @field_validator('technologies', mode='before')
    @classmethod
    def parse_technologies(cls, v):
        # 数据库中技术栈以JSON字符串存储，读取时自动反序列化
        if isinstance(v, str):
            return json.loads(v) if v else []
        return v if v is not None else []


class PortfolioItemCreate(PortfolioItemBase):
    title: str
    slug: str


class PortfolioItemUpdate(BaseModel):
    title: Optional[str] = None
    slug: Optional[str] = None
    description: Optional[str] = None
    cover_image: Optional[str] = None
    demo_url: Optional[str] = None
    github_url: Optional[str] = None
    technologies: Optional[List[str]] = None
    start_date: Optional[date] = None
    end_date: Optional[date] = None
    status: Optional[str] = None
    is_featured: Optional[bool] = None
    sort_order: Optional[int] = None


class PortfolioItemInDBBase(PortfolioItemBase):
    id: UUID
    created_at: datetime
    updated_at: Optional[datetime] = None

    @field_serializer('id')
    def serialize_id(self, value: UUID) -> str:
        return str(value)

    model_config = {'from_attributes': True}


class PortfolioItem(PortfolioItemInDBBase):
    pass


class Portfolio(PortfolioItemInDBBase):
    pass


class PortfolioCreate(PortfolioItemBase):
    title: str
    slug: str


class PortfolioUpdate(BaseModel):
    title: Optional[str] = None
    slug: Optional[str] = None
    description: Optional[str] = None
    cover_image: Optional[str] = None
    demo_url: Optional[str] = None
    github_url: Optional[str] = None
    technologies: Optional[List[str]] = None
    start_date: Optional[date] = None
    end_date: Optional[date] = None
    status: Optional[str] = None
    is_featured: Optional[bool] = None
    sort_order: Optional[int] = None